### PR TITLE
add copy to SearchControl (allow multiple) + support paste for ImplementedByAll and EntityStrip

### PR DESCRIPTION
**Pull request by mehdy-karimpour**

### CODE REVIEW

File: Signum.Engine.Extensions/Files/FilePathEmbeddedLogic.cs [hunk 1]
```diff
@@ -23,13 +23,22 @@ public static void Start(SchemaBuilder sb)
...
             {
-                if(efp.BinaryFile != null) //First time
+                if (efp.BinaryFile != null) //First time
                 {
                     var task = efp.SaveFileAsync();
                     Transaction.PreRealCommit += data =>
                     {
                         var a = efp; //For debugging
-                        task.Wait();
+                        if (task.Status == TaskStatus.WaitingForActivation)
```
Comment (olmobrutall) [inline Signum.Engine.Extensions/Files/FilePathEmbeddedLogic.cs:32]: This is a hack. We need a better solution

File: Signum.React/Scripts/Lines/EntityBase.tsx [hunk 2]
```diff
@@ -264,6 +266,23 @@ export class EntityBaseController<P extends EntityBaseProps> extends LineBaseCon
...
+    event.preventDefault();
+
+    navigator.clipboard.readText()
+      .then(text => {
+        const lites = parseLiteList(text);
+        if (lites.length == 0)
+          return;
+
+        const lite = lites[0];
+        Navigator.API.fillToStrings(lite)
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/EntityBase.tsx:280]: Return missing

File: Signum.React/Scripts/Lines/EntityBase.tsx [hunk 3]
```diff
@@ -264,6 +267,42 @@ export class EntityBaseController<P extends EntityBaseProps> extends LineBaseCon
...
+        if (lites.length == 0)
+          return;
+
+        const ti = this.props.type!.name == IsByAll ? getTypeInfo(lites[0].EntityType) : getTypeInfos(this.props.type!).singleOrNull(ti => ti.name == lites[0].EntityType);
+        if (!ti)
+          return;
+
+        const promise = lites.length == 1 ? Promise.resolve(lites[0]) : 
+          Navigator.API.fillToStrings(...lites)
+            .then(() => SelectorModal.chooseLite(new Type<Entity>(lites[0].EntityType), lites));
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/EntityBase.tsx:286]: There are three problems here: 
* you are not calling fillToStrings for the length == 1 case
* SelectorModal.chooseLite is silent if there is only one lite, no need to split the cases.
* SelectorModal.chooseLite should take a Type<T> | TypeInfo | string as first parameter

Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/EntityBase.tsx]: Excellent catches 👍 , fixed

File: Signum.React/Scripts/Lines/EntityBase.tsx (hunk 2 above)
Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/EntityBase.tsx]: Perfect 👍

File: Signum.Engine.Extensions/Files/FilePathEmbeddedLogic.cs (hunk 1 above)
Comment (mehdy-karimpour) [reply Signum.Engine.Extensions/Files/FilePathEmbeddedLogic.cs]: Temporarily removed to accept PR 👍

File: Signum.React/Scripts/Lines/AutoCompleteConfig.tsx [hunk 4]
```diff
@@ -188,21 +186,32 @@ export class FindOptionsAutocompleteConfig implements AutocompleteConfig<ResultR
...
 
-    if (FindOptionsAutocompleteConfig.liteKeyRegEx.test(subStr)) {
-      const lite = parseLite(subStr);
-      const tis = getTypeInfos(qd.columns["Entity"].type);
-      const ti = tis.singleOrNull(ti => ti.name == lite.EntityType);
+    const lites = parseLiteList(subStr);
+    if (lites.length > 0) {
+      const type = qd.columns["Entity"].type;
+      const tis = getTypeInfos(type);
+      const ti = tis.singleOrNull(ti => ti.name == lites[0].EntityType);
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/AutoCompleteConfig.tsx:193]: I don't understand this complexity. Why not filter all the `lites` by having one of the types in `tis` and then add a `Entity` `EqualsTo` / `IsIn` with the list.

File: Signum.React/Scripts/Lines/EntityBase.tsx [hunk 5]
```diff
@@ -264,6 +267,40 @@ export class EntityBaseController<P extends EntityBaseProps> extends LineBaseCon
...
+
+    event.preventDefault();
+
+    navigator.clipboard.readText()
+      .then(text => {
+        const lites = parseLiteList(text);
+        if (lites.length == 0)
+          return;
+
+        const ti = this.props.type!.name == IsByAll ? getTypeInfo(lites[0].EntityType) : getTypeInfos(this.props.type!).singleOrNull(ti => ti.name == lites[0].EntityType);
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/EntityBase.tsx:280]: again better to filter the lites that are compatible. Maybe you pasting a list of apples and oranged in a orange field.

File: Signum.React/Scripts/Lines/EntityBase.tsx [hunk 6]
```diff
@@ -277,6 +314,19 @@ export class EntityBaseController<P extends EntityBaseProps> extends LineBaseCon
     );
   }
 
+  renderPasteButton(btn: boolean) {
+    if (this.props.ctx.readOnly || !this.props.type || (this.props.type.name != IsByAll && !this.props.type.isCollection))
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/EntityBase.tsx:318]: EntityBase should have a `paste` property which default value is set to `this.props.type.name == IsByAll` in `getDefaultProps`, similar to how the `create`/`find`/`view` properties works.

File: Signum.React/Scripts/Lines/EntityListBase.tsx [hunk 7]
```diff
@@ -159,6 +162,33 @@ export abstract class EntityListBaseController<T extends EntityListBaseProps> ex
     this.setValue(list);
   }
 
+  handlePasteClick = (event: React.SyntheticEvent<any>) => {
+
+    event.preventDefault();
+
+    navigator.clipboard.readText()
+      .then(text => {
+        const lites = parseLiteList(text);
```
Comment (olmobrutall) [inline Signum.React/Scripts/Lines/EntityListBase.tsx:171]: again filtering the lites

**Review by olmobrutall — COMMENTED** [review]
Just a few issues more :)

File: Signum.React/Scripts/Lines/AutoCompleteConfig.tsx (hunk 4 above)
Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/AutoCompleteConfig.tsx]: done 👍

File: Signum.React/Scripts/Lines/EntityBase.tsx (hunk 5 above)
Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/EntityBase.tsx]: done 👍

File: Signum.React/Scripts/Lines/EntityBase.tsx (hunk 6 above)
Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/EntityBase.tsx]: done 👍

File: Signum.React/Scripts/Lines/EntityListBase.tsx (hunk 7 above)
Comment (mehdy-karimpour) [reply Signum.React/Scripts/Lines/EntityListBase.tsx]: done 👍

**Comment by olmobrutall** [comment]
Thanks for all the great work 👍

**Comment by mehdy-karimpour** [comment]
You`re welcome and thanks for your help 👍

**Pull request merged**